File: 404.sum-of-left-leaves.js
/*
 * @lc app=leetcode id=404 lang=javascript
 *
 * [404] Sum of Left Leaves
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
const sumOfLeftLeaves = function (root)
{
    const left = root.left;
    const right = root.right;
    if (left === null && right === null)
    {
        return 0;
    }

    let sum = 0;

    if (left !== null)
    {
        if (left.left === null && left.right === null)
        {
            sum += left.val;
        }
        else
        {
            sum += sumOfLeftLeaves(left);
        }
    }

    if (right !== null)
    {
        sum += sumOfLeftLeaves(right);
    }

    return sum;
};
// @lc code=end

File: src/2023-2024/404.sum-of-left-leaves.cpp
/*
 * @lc app=leetcode id=404 lang=cpp
 *
 * [404] Sum of Left Leaves
 */
struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right)
      : val(x), left(left), right(right) {}
};
// @lc code=start

class Solution {
 public:
  int sumOfLeftLeaves(TreeNode* root) {
    if (root == nullptr) {
      return 0;
    }
    int leftChildSum =
        root->left == nullptr
            ? 0
            : (isLeafNode(root->left) ? root->left->val
                                      : sumOfLeftLeaves(root->left));
    int rightChildSum = sumOfLeftLeaves(root->right);

    return leftChildSum + rightChildSum;
  }

  inline bool isLeafNode(const TreeNode* node) {
    return node->left == nullptr && node->right == nullptr;
  }
};
// @lc code=end
